packages/material-ui-lab/src/TreeItem/TreeItem.test.js

```diff
@@ -436,6 +436,40 @@ describe('<TreeItem />', () => {
           expect(getByTestId('two')).to.have.focus;
         });
 
+        it('moves focus to a sibling node after conditional rendering', () => {
```

Comment:
I don't think this is the correct semantic. In the DOM elements simply get blurred when they're removed. And I think we don't move focus in the other widget components if a child is removed.

Could you elaborate a bit more why you think we should move focus to another child when an element is removed?

Reply:
Yes, that test name choice is not the best.
it('moves focus to a sibling node and works with conditional rendering') or something similar.

Reply:
It is a test that expected behaviour ( which already has tests ) still occurs when we have a dynamic tree.

Without this commit arrow down does not work with a dynamic tree.

Reply:
Sure but this is very unusual behavior. Normally focus is not removed when an element is removed. I don't think it's in general the right approach to focus the next sibling if an item gets removed. 

I guess this is why moving focus with `aria-activedescendant` is better since you won't encounter these issues.

Either way: If you remove a focus from an item then you can always focus the next sibling in your app. I would rather discuss this in an issue that starts with a concrete use case.

Reply:
The concrete use case is https://github.com/mui-org/material-ui/issues/20140.  This is the same as the failing test in @joshwooding draft [pull request](https://github.com/mui-org/material-ui/pull/19973/files).

You misunderstand what is happening here ( I assume because of the test description ).  This has nothing to do with the removal of items.  It is that when the tree is dynamic the building of visible nodes is incorrect and that arrow down ( and no doubt other navigation ) does not work properly.  In the example we focus the wrong item and in other circumstances you cannot focus at all with the down arrow !
https://codesandbox.io/s/silly-sound-r3x25

Reply:
Could you then add a test that explains what is happening?

Do we agree that removing an item should not move focus but just blur that item?

Reply:
The test is it(''moves focus to a sibling node after conditional rendering'...  which should have been worded [describe('down arrow interaction', () => {]  it(''moves focus to a sibling node **works with** conditional rendering'.

In the test the tree does not have focus.  The tree is updated.
We focus.
We navigate with the down arrow. **Behaviour is incorrect**
( Nothing is removed )

As for what the behaviour should be when we remove an item.

The tree needs to be navigatable by the keyboard.  If the tree has items after removal then given the way that focus is achieved currently it should have one li with tabindex=0.  Currently if the tree updates and the tree item that is the tabbable node id has been removed then there is no code to then set tabbable to one of the other tree items ( the first with the current implementation.)  We need to add focus management to our clean up tasks when a node is removed. 

Personally I propose a different focus strategy.
None of the tree items have tabIndex=0 until they are :
a) clicked or 
b) the tree itself receives focus.

b) The ul is given tabIndex and handles onFocus - it then can look at **selected**.
If there are selected then it focuses ( calls focus ) the selected ( single-select ) or the first selected ( multi-select) and if no selected then focuses the first tree item as per https://www.w3.org/TR/wai-aria-practices/#TreeView

> When a single-select tree receives focus:
If none of the nodes are selected before the tree receives focus, focus is set on the first node.
If a node is selected before the tree receives focus, focus is set on the selected node.
When a multi-select tree receives focus:
If none of the nodes are selected before the tree receives focus, focus is set on the first node.
If one or more nodes are selected before the tree receives focus, focus is set on the first selected node.

In addition to focusing a tree item it then needs to remove itself from tabbing via state.

Once the tree view has focus on one of the items then navigation works as is.

When focus leaves the tree the tree is then made tabbable again.

Reply:
>  which should have been worded

This is your PR. Please change anything that doesn't seem right to you. It seems we agree that the current description is incorrect.

Also "moves focus" implies focus is moved between two elements. It seems like you want to focus a specific item once the item gets rendered. How would this work if focus was on a completely different part of the UI?

Reply:
Ok I have changed the test name and swapped getByText.

The test name was wrong as the test is that ArrowDown moves focus to a child node.

It is the same test as the one above it that is not dynamic.

```
//tree does not have focus.

//force the tree to update its items 
expect(getByTestId('one')).to.not.be.null;
fireEvent.click(getByTestId('button'));
expect(queryByTestId('one')).to.be.null;
fireEvent.click(getByTestId('button'));
expect(getByTestId('one')).to.not.be.null;

//give focus to tree
getByTestId('one').focus(); 
//"down arrow interaction" "moves focus...." 
fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });

expect(getByTestId('two')).to.have.focus;
```

I don't want to focus a specific item when it gets rendered, I want the focus to be moved from the one that has focus to the one that should receive focus when the down arrow is pressed.

Reply:
@eps1lon now that the test name has been corrected can we correct this bug ?